File: nc_150/trees/level_order_traversal.cpp
// the idea is to make use of queue and put the nodes level by level
// keep count of the nodes at a particular level
// TC = O(n)


vector<vector<int>> levelOrder(TreeNode* root) {
    vector<vector<int>>ans ;
    queue<TreeNode*>q ;

    if (root) q.push(root);
    while(!q.empty()){
        int n = q.size();
        vector<int>temp ;
        while(n--){
            TreeNode* node = q.front();
            q.pop();
            temp.push_back(node -> val);
            if (node -> left) q.push(node -> left);
            if (node -> right) q.push(node -> right);
        }
        ans.push_back(temp);
    }
    return ans ;
}

File: nc_150/trees/go/level_order_traversal.go
// the idea is to store queue len at each level before pushing the child nodes onto the queue
// this way we can iterate level by level

func levelOrder(root *TreeNode) [][]int {
	var res [][]int
	if root == nil {
		return res
	}

	q := []*TreeNode{root}

	for len(q) > 0 {
		level_cnt := len(q)
		var temp []int
		for i := 0; i < level_cnt; i++ {
			node := q[0]
			temp = append(temp, node.Val)
			q = q[1:]

			if node.Left != nil {
				q = append(q, node.Left)
			}
			if node.Right != nil {
				q = append(q, node.Right)
			}
		}
		res = append(res, temp)
	}
	return res
}